### PR TITLE
Multiple roots and federation

### LINKED ISSUE #19: Add basic federation where you can add multiple roots to your trust bundle
Also a useful feature for root certs rotation.

### PR DESCRIPTION
**Pull request by maraino**
### Description
This PR adds support for federated roots, and multiple roots for root rotation. Fixes #19

Federated roots are a list of public certificates that a client can include and trust, a new endpoint (`/federation`) has been added that returns the server roots as well as the federated ones.

The CA now supports one or multiple roots in the `root` property of the JSON, that property is able to unmarshal both strings and arrays, the latter must be always used for multiple roots. The SDK is able to update the root certificates on each renew, and this feature can be used for root rotation following this process:
1. CA is configured with rootA.
2. BootstrapServer/BootstrapClient are created, accepting only rootA.
3. CA is updated and adds rootB, rootA and rootB are accepted.
4. BootstrapServer/BootstrapClient do a renew and include rootB into the trusted pools
5. CA can now create new intermediates and remove rootA.
6. BootstrapServer/BootstrapClient will able to renew with rootB.

A known issue is that at this moment rootA won't be removed, but a new issue with a strategy to support that has been created.

### CODE REVIEW

File: api/api_test.go [hunk 1]
```diff
@@ -443,6 +445,20 @@ func (m *mockAuthority) GetEncryptedKey(kid string) (string, error) {
 	return m.ret1.(string), m.err
 }
 
+func (m *mockAuthority) GetRoots() ([]*x509.Certificate, error) {
+	if m.getFederation != nil {
```
Comment (dopey) [inline api/api_test.go]: m.getRoots?

Comment (maraino) [reply api/api_test.go]: Fixed 8252608

File: ca/client.go [hunk 2]
```diff
@@ -413,6 +399,42 @@ func (c *Client) ProvisionerKey(kid string) (*api.ProvisionerKeyResponse, error)
...
+func (c *Client) Roots() (*api.RootsResponse, error) {
+	u := c.endpoint.ResolveReference(&url.URL{Path: "/roots"})
+	resp, err := c.client.Get(u.String())
+	if err != nil {
+		return nil, errors.Wrapf(err, "client GET %s failed", u)
+	}
+	if resp.StatusCode >= 400 {
+		return nil, readError(resp.Body)
+	}
+	var federation api.RootsResponse
```
Comment (dopey) [inline ca/client.go]: var roots api.RootsResponse?

Comment (maraino) [reply ca/client.go]: Fixed in dbd1bf1

File: authority/root.go [hunk 3]
```diff
@@ -17,13 +17,39 @@ func (a *Authority) Root(sum string) (*x509.Certificate, error) {
...
 	return crt, nil
 }
 
 // GetRootCertificate returns the server root certificate.
 func (a *Authority) GetRootCertificate() *x509.Certificate {
-	return a.rootX509Crt
+	return a.rootX509Certs[0]
+}
+
+// GetRootCertificates returns the server root certificates.
```
Comment (dopey) [inline authority/root.go]: Here we have two methods in the public Authority API that return the same thing. The explanation behind this duplication is that one of these methods is used by the CA API while the other is used by the SDK API. It is possible that the implementations for the methods change (e.g. new conditions are added to one and not the other) in the future.

File: authority/types.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+// First returns the first element of a multiString. It will return an empty
+// string if the multistring is empty.
+func (s multiString) First() string {
+	if len(s) > 0 {
+		return s[0]
+	}
+	return ""
+}
+
+// Empties returns `true` if any string in the array is empty.
```
Comment (dopey) [inline authority/types.go]: Consider changing this to `HasEmpties`? NitPicky and I'm not entirely sure that the name change has any effect on clarity. It's just that `Has...` is a more commonly used name for this sort of thing.

Comment (maraino) [reply authority/types.go]: Fixed in 6e62007

File: authority/root.go (hunk 3 above)
Comment (maraino) [reply authority/root.go]: Fixed in e8ac3f4

**Pull request merged**